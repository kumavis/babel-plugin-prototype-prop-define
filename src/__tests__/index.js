import path from 'path'
import pluginTester from 'babel-plugin-tester'
import plugin from '../'

const projectRoot = path.join(__dirname, '../../')

expect.addSnapshotSerializer({
  print(val) {
    return val.split(projectRoot).join('<PROJECT_ROOT>/')
  },
  test(val) {
    return typeof val === 'string'
  },
})

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {filename: __filename},
  tests: {
    'should ignore': `

      const x = {}
      x.a = () => true
  
    `,
    'on object': `

      const x = {}
      x.toString = () => true
    
    `,
    'on prototype': `

      function MyClass () {}
      MyClass.prototype.toString = () => true

    `,
    'on array': `

      const x = []
      x.splice = () => true

    `,
    'on function': `

      const x = () => {}
      x.bind = () => true

    `,
    'ignore alternate assigment expressions': `

      const x = {}
      x.toString += 1
    
    `,
    // 'key as string': `

    //   const x = {}
    //   x["toString"] = () => true

    // `,
  },
})
