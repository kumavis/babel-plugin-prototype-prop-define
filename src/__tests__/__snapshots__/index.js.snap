// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prototype-prop-define dynamic key: dynamic key 1`] = `

const x = {}
;['toString'].forEach(key => {
  x[key] = () => true
})

      ↓ ↓ ↓ ↓ ↓ ↓

const x = {};
['toString'].forEach(key => {
  (function (parent, key, value) {
    ["constructor", "__defineGetter__", "__defineSetter__", "hasOwnProperty", "__lookupGetter__", "__lookupSetter__", "isPrototypeOf", "propertyIsEnumerable", "toString", "valueOf", "__proto__", "toLocaleString", "length", "concat", "find", "findIndex", "pop", "push", "shift", "unshift", "slice", "splice", "includes", "indexOf", "keys", "entries", "forEach", "filter", "map", "every", "some", "reduce", "reduceRight", "join", "reverse", "sort", "lastIndexOf", "copyWithin", "fill", "values", "name", "arguments", "caller", "apply", "bind", "call", "message", Symbol.iterator, Symbol.unscopables, Symbol.hasInstance, Symbol.asyncIterator, Symbol.toStringTag].includes(key) ? Object.defineProperty(parent, key, {
      value: value,
      writable: true,
      enumerable: true,
      configurable: true
    }) : parent[key] = value;
    return value;
  })(x, key, () => true);
});

`;

exports[`prototype-prop-define ignore alternate assigment expressions: ignore alternate assigment expressions 1`] = `

const x = {}
x.toString += 1

      ↓ ↓ ↓ ↓ ↓ ↓

const x = {};
x.toString += 1;

`;

exports[`prototype-prop-define key as string: key as string 1`] = `

const x = {}
x["toString"] = () => true

      ↓ ↓ ↓ ↓ ↓ ↓

const x = {};

(function (parent, key, value) {
  Object.defineProperty(parent, key, {
    value: value,
    writable: true,
    enumerable: true,
    configurable: true
  })
  return value;
})(x, "toString", () => true);

`;

exports[`prototype-prop-define on array: on array 1`] = `

const x = []
x.splice = () => true

      ↓ ↓ ↓ ↓ ↓ ↓

const x = [];

(function (parent, key, value) {
  Object.defineProperty(parent, key, {
    value: value,
    writable: true,
    enumerable: true,
    configurable: true
  })
  return value;
})(x, "splice", () => true);

`;

exports[`prototype-prop-define on function: on function 1`] = `

const x = () => {}
x.bind = () => true

      ↓ ↓ ↓ ↓ ↓ ↓

const x = () => {};

(function (parent, key, value) {
  Object.defineProperty(parent, key, {
    value: value,
    writable: true,
    enumerable: true,
    configurable: true
  })
  return value;
})(x, "bind", () => true);

`;

exports[`prototype-prop-define on object: on object 1`] = `

const x = {}
x.toString = () => true

      ↓ ↓ ↓ ↓ ↓ ↓

const x = {};

(function (parent, key, value) {
  Object.defineProperty(parent, key, {
    value: value,
    writable: true,
    enumerable: true,
    configurable: true
  })
  return value;
})(x, "toString", () => true);

`;

exports[`prototype-prop-define on prototype: on prototype 1`] = `

function MyClass () {}
MyClass.prototype.toString = () => true

      ↓ ↓ ↓ ↓ ↓ ↓

function MyClass() {}

(function (parent, key, value) {
  Object.defineProperty(parent, key, {
    value: value,
    writable: true,
    enumerable: true,
    configurable: true
  })
  return value;
})(MyClass.prototype, "toString", () => true);

`;

exports[`prototype-prop-define should ignore: should ignore 1`] = `

const x = {}
x.a = () => true

      ↓ ↓ ↓ ↓ ↓ ↓

const x = {};

x.a = () => true;

`;

exports[`prototype-prop-define symbol assignment: symbol assignment 1`] = `

const x = {}
x[Symbol.iterator] = () => true

      ↓ ↓ ↓ ↓ ↓ ↓

const x = {};

(function (parent, key, value) {
  Object.defineProperty(parent, key, {
    value: value,
    writable: true,
    enumerable: true,
    configurable: true
  })
  return value;
})(x, Symbol.iterator, () => true);

`;
